packages/preview-middleware-client/src/adp/control-utils.ts

```diff
@@ -0,0 +1,484 @@
+/* eslint-disable consistent-return */
+import Utils from 'sap/ui/fl/Utils';
+import DataType from 'sap/ui/base/DataType';
+import type ElementOverlay from 'sap/ui/dt/ElementOverlay';
+import type ManagedObject from 'sap/ui/base/ManagedObject';
+import type { MetadataOptions } from 'sap/ui/base/ManagedObject';
+
+declare const jQuery: {
+    sap: {
+        getObject: (object: unknown) => { [key: string]: string };
+    };
+};
+
+export interface BuiltRuntimeControl {
+    id: string;
+    type: string;
+    properties: Properties[];
+    name: string;
+}
+
+export interface Properties {
+    type: string;
+    editor: string;
+    name: string;
+    readableName: string;
+    value: unknown | boolean;
+    isEnabled: boolean;
+    documentation: object;
+    options?: { key: string; text: string }[];
+    isIcon?: boolean;
+}
+
+export interface AnalyzedType {
+    primitiveType: string;
+    ui5Type: string | null;
+    enumValues: { [key: string]: string } | null;
+    isArray: boolean;
+}
+
+export type ControlNewData = {
+    id: any;
+    name: any;
+    newValue: any;
+};
+
+export type MetadataOptionsProperty = MetadataOptions.Property & {
+    name: string;
+    getType(): {
+        getName(): string;
+    };
+};
+
+/**
+ * Handles calling control specific functions for retrieving control data
+ */
+export default class ControlUtils {
+    /**
+     * Analyzed type
+     */
+    private static analyzedType: AnalyzedType;
+
+    /**
+     * Returns ManagedObject runtime control
+     *
+     * @param overlayControl Overlay
+     * @returns {ManagedObject} Managed Object instance
+     */
+    public static getRuntimeControl(overlayControl: ElementOverlay): ManagedObject {
+        let runtimeControl;
+        if (overlayControl.getElementInstance) {
+            runtimeControl = overlayControl.getElementInstance();
+        } else {
+            runtimeControl = overlayControl.getElement();
+        }
+        return runtimeControl;
+    }
+
+    /**
+     * Returns control aggregation names in an array
+     *
+     * @param control Managed Object runtime controll
+     * @param name Aggregation name
+     * @returns Array of control aggregations
+     */
+    public static getControlAggregationByName(
+        control: ManagedObject & { __calledJSONKeys?: boolean; [key: string]: any },
+        name: string
+    ) {
+        let result = [];
+        const aggregation = (control ? control.getMetadata().getAllAggregations() : {})[name] as unknown as object & {
+            _sGetter: string;
+        };
+
+        if (aggregation) {
+            if (!aggregation._sGetter && !control.__calledJSONKeys) {
+                (control.getMetadata() as any).getJSONKeys();
```

Comment:
no `any`. Additionally, this whole function needs to be revisited. Accessing private variables and attaching new ones does not look like maintainable software.